core/ir/decode_api.h

```diff
@@ -57,6 +57,17 @@ DR_API
 byte *
 decode_eflags_usage(void *drcontext, byte *pc, uint *usage, dr_opnd_query_flags_t flags);
 
+#ifdef AARCH64
+DR_API
+/**
+ * Attempts to decode the instruction at address \p pc to determine if the
+ * instruction is supported by the decoder.
+ * Returns false on decoding an invalid or unsupported instruction.
```

Comment:
This does not seem a necessary interface addition: an invalid instruction will already return NULL and set OP_INVALID; an unsupported instruction will have OP_xx.

Is the instruction causing the test failure OP_xx?

Reply:
The problem is with both unsupported and invalid ops, although I think the aarch64 decoder will classify both as OP_xx. In [decode_common](https://github.com/DynamoRIO/dynamorio/blob/0496c7fe38a1fb69c89a94ea989cd1da07e515c7/core/ir/aarch64/codec.c#L4524) if the decoder returns false (due to an invalid or an unsupported inst)  then it will still populate various fields of the instr and call opnd_create_reg. This is fine for just an unsupported instruction that is well-formed, but with a completely random and invalid instruction, this code could cause the assertion you reported.

Reply:
I see: so that is a bug in the decoder.  But how does it get an invalid reg value: wouldn't it use 4 bits and any 4-bit value is ok, or is this reg 31 not being ok somewhere?  It seems best to fix this in the decoder: is it a small fix?

Reply:
I couldn't reproduce the issue and without a callstack it is unclear who called `opnd_create_reg` leading to the assertion. However, we know that it is during generating the random code, and it should be when decoding the random inst. So, it is inside `decode_common`, either in the call to the [decoder](https://github.com/DynamoRIO/dynamorio/blob/0496c7fe38a1fb69c89a94ea989cd1da07e515c7/core/ir/aarch64/codec.c#L4524) or in the code immediately after. The decoder calls various functions to create reg operands that are defined in [codec.c](https://github.com/DynamoRIO/dynamorio/blob/980099e42d7ce0cec55800aeefb0357d9274e98b/core/ir/aarch64/codec.c). I checked all the uses of `opnd_create_reg` in `codec.c` and none of them seem to be able to produce a DR_REG_INVALID or > DR_REG_LAST_ENUM  . The only weird case is in the `decode_common` where the decoder returns false and  instead of using `extract_uint(enc, pos, n)` as in all other places it directly does the shift and the mask without forcing an uint. It does not seem likely but is it possible that the result of this shift and mask could be interpreted as a 5-bit negative number or can some subcomputation be interpreted as signed? Probably worth changing this part in any case to conform with the norm in codec.c.